src/passes/TypeGeneralizing.cpp

```diff
@@ -323,81 +394,470 @@ struct TransferFn : OverriddenVisitor<TransferFn> {
+
+  void visitCallRef(CallRef* curr) {
+    auto sigType = curr->target->type.getHeapType();
+    if (sigType.isBottom()) {
+      // This will be emitted as an unreachable, so impose no requirements on
+      // the arguments, but do require that the target continue to have bottom
+      // type.
+      clearStack();
+      push(Type(HeapType::nofunc, Nullable));
+      return;
+    }
+
+    auto sig = sigType.getSignature();
+    auto numParams = sig.params.size();
+    std::optional<Type> resultReq;
+    if (sig.results.isRef()) {
+      resultReq = pop();
+    }
+
+    // We have a choice here: We can either try to generalize the type of the
+    // incoming function reference or the type of the incoming function
+    // arguments. Because function parameters are contravariant, generalizing
+    // the function type inhibits generalizing the arguments and vice versa.
+    // Attempt to split the difference by generalizing the function type only as
+    // much as we can without imposing stronger requirements on the arguments.
+    auto targetReq = sigType;
+    while (true) {
+      auto candidateReq = targetReq.getDeclaredSuperType();
+      if (!candidateReq) {
+        // There is no more general type we can require.
+        break;
+      }
+
+      auto candidateSig = candidateReq->getSignature();
+
+      if (resultReq && *resultReq != candidateSig.results &&
+          Type::isSubType(*resultReq, candidateSig.results)) {
+        // Generalizing further would violate the requirement on the result
+        // type.
+        break;
+      }
+
+      for (size_t i = 0; i < numParams; ++i) {
+        if (candidateSig.params[i] != sig.params[i]) {
+          // Generalizing further would restrict how much we could generalize
+          // this argument, so we choose not to generalize futher.
+          // TODO: Experiment with making the opposite choice.
+          goto done;
+        }
+      }
+
+      // We can generalize.
+      targetReq = *candidateReq;
+    }
+  done:
+
+    // Push the new requirements for the parameters.
+    auto targetSig = targetReq.getSignature();
+    for (auto param : targetSig.params) {
+      if (param.isRef()) {
+        push(param);
+      }
+    }
+    // The new requirement for the call target.
+    push(Type(targetReq, Nullable));
+  }
+
+  void visitRefTest(RefTest* curr) {
+    // Do not require anything of the input.
+    push(Type::none);
+  }
+
+  void visitRefCast(RefCast* curr) {
+    // In principle, we do not have to require anything of the input. However,
+    // if our goal is to eliminate casts, then we want the input to be a subtype
+    // of the output, so we don't want the input to be generalized too much.
+    // Propagate the requirement on our output to our input if we can, and
+    // otherwise keep the input type unchanged. Unfortunately we cannot fall
+    // back to requiring nothing of our input because that would not be
+    // monotonic.
```

Comment:
I don't follow this. No matter how much we generalize the input, we can generalize the cast in lockstep, can't we? I don't see a situation where we end up with the input "too generalized".

Reply:
Monotonicity here means that the tighter the constraint on the output to the cast, the tighter the constraint we must propagate to the input to the cast. If the constraint on the output gets tighter such that we realize we won't be able to remove the cast, we cannot then loosen the constraint on the input.

Reply:
To check if I follow, does "tighter" means "more refined" here?

Reply:
Yes, that's right.

Reply:
Ok, even with that I can't seem to follow this. Isn't monotonicity a property of each location? That is, as we perform work, we must change the type of this cast monotonically, and separately the type of the child. Concrete example:

```wat
(use-A ;; uses type $A
  (ref.cast $B
    (block
      (ref.cast $C
        ..
```
The outer cast (of `$B`) is informed that we only need an `$A`, so we monotonically change its requirements from the initial top type ("no requirement") to $A. We can still do nothing at all for the block that is its child; it began as top and remains as top, which is also monotonic.

Continuing on, since the block has top type (no requirements reach it) so does the inner cast, so we can modify the casts like this:
```wat
(use-A ;; uses type $A
  (ref.cast $A
    (block
      (ref.cast null any
        ..
```
Then OptimizeInstructions can remove both casts.

Did I miss a place where monotonicity was lost here?

Reply:
No, monotonicity is a property of the transfer function operating on the single lattice encompassing all program locations. If the input to the transfer function gets higher up the lattice because our constraint on one program location becomes tighter, then if the transfer function responded by loosening the constraint on another program location, then the transfer function would not be monotonic.

Your example doesn't do anything wrong as far as I can tell.

The non-monotonic implementation this comment is warning against is this:

```c++
auto req = pop();
if (req != curr->ref->type && Type::isSubType(req, curr->ref->type) {
  // We will not be able to remove this cast anyway, so we are free to
  // generalize the input.
  push(Type::none);
} else {
  // We still might be able to remove the cast, but only if we don't generalize
  // the input any further than the output.
  push(req);
}
```

In this implementation, as the popped requirement on the output `req` gets more refined, the pushed requirement on the input can relax from `req` to `Type::none`. That implementation might allow more previous casts to be optimized out, but since it is not monotonic, it might also cause the analysis to infinitely loop.

Reply:
Thanks, that code makes it clearer for me. To make sure I follow, that would be non-monotonic in the sense I mentioned, correct? Those pushes could go from pushing increasingly refined types `$A, $B, $C`, going downwards more and more, to suddenly pushing `none` which is at the very top.

Now that I (hopefully) follow that, I'm still not clear on the code below. As I wrote above, I would just push `none` unconditionally here. It seems like you push the LUB in an attempt to avoid a situation where you lose the ability to remove a cast, but I can't see how that can occur: if in fact we become not-refined-enough for a cast, that suggests to me that we removed another cast somewhere else, to achieve that.

Also, it seems worrying to me to try to find some tricky compromise in between always generalizing and sometimes generalizing. It seems simpler to always generalize, period, and then if necessary to do some sort of followup operation to improve things if we need to (but as I said in the previous paragraph, I don't see why we'd need even that).

Is there a simple situation I'm missing here?

Reply:
Yes, sounds like you're following correctly 👍 

Here's a trivial example input where unconditionally pushing none would inhibit a cast optimization:

```
(func $foo (param $i i32) (result i31ref)
 (local $x i31ref)
 (ref.cast i31ref
  (local.get $x)
 )
)
```
If the cast pushed `Type::none` as its input requirement, then we would end up generalizing `$x` to `anyref` and would no longer be able to optimize out the cast.

A less trivial example might involve a struct.get that feeds into a cast. If we allow the cast input to be generalized too much, then the struct reference we pass to the struct.get might become general enough that the accessed field is no longer a subtype of the cast's required output, so we wouldn't be able to eliminate the cast any more.

Reply:
The trivial example is correct, but I'm not sure it's what we should aim to solve. `--type-generalizing --local-refining` turns that local into `nullref` anyhow.

The less trivial example is more plausible, but it could easily go the other way - if we generalize the struct reference enough we might well remove a cast there, as I said before. I don't have data on this, but my intuition is that this is a zero-sum game (an attempt to improve in one place worsens in another).

The attempt to strike a middle ground is also more complicated. Unless I'm mistaken, we don't have good data pointing to either option, so I'm in favor of the simple one, that is, to maximally generalize.

Reply:
Yes, fair enough. This is a policy decision similar to the one we make for ref.cast. My intuition is that it is better to try to optimize the cast we know about than the one we don't know about, but that may very well turn out to be incorrect. I'll simplify the code here and add a TODO to experiment with it.

FWIW, an advanced version of this analysis could find not just a single constraint for each program location, but rather a set of constraints on program locations for each cast that would allow that cast to be optimized out. We could then feed that system of constraints to a SMT solver to maximize the number of eliminated casts. I doubt this is something we would want to spend time on, but it might make for an interesting intern project some day.